pkg/shared/kvs/jetstream/kv_store.go

```diff
@@ -289,22 +290,28 @@ retryLoop:
 
 	}
 
-	for _, key := range keys {
+	for key := range keyLister.Keys() {
 		value, err = jss.kv.Get(key)
 		for err != nil {
+			// keys can be deleted when the previous vertex pod is deleted/restarted.
+			if errors.Is(err, nats.ErrKeyNotFound) {
+				break
+			}
 			select {
 			case <-jss.ctx.Done():
 				return time.Time{}
 			default:
-				jss.log.Errorw("Failed to get value", zap.String("watcher", jss.GetKVName()), zap.Error(err))
-				value, err = jss.kv.Get(key)
+				jss.log.Errorw("Failed to get value", zap.String("watcher", jss.GetKVName()), zap.String("key", key), zap.Error(err))
 				time.Sleep(100 * time.Millisecond)
+				value, err = jss.kv.Get(key)
 			}
 		}
 		if value.Created().After(lastUpdate) {
 			lastUpdate = value.Created()
 		}
 	}
+
+	_ = keyLister.Stop()
```

Comment:
will we end up in double stopping because KeyLister closes stop if all is read?

Reply:
Shall we use `defer`?